[9.x] Add Str::highlight()

BramVanBerkel opened the pull request:
After seeing #41000, I got inspired to also add the highlight feature, the original idea also coming from [Rails](https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-highlight)

This pull requests adds the ability to easily highlight words in a string:
```php
public static function highlight($text, $highlight, $tag = 'mark')
```

Usage:
```php
Str::highlight('Laravel, The PHP Framework for Web Artisans', 'Laravel');
// <mark>Laravel</mark>, The PHP Framework for Web Artisans

Str::highlight('Laravel, The PHP Framework for Web Artisans', ['Laravel', 'Framework', 'Artisans']);
// <mark>Laravel</mark>, The PHP <mark>Framework</mark> for Web <mark>Artisans</mark>
```
Or with custom tags:
```php
Str::highlight('Laravel, The PHP Framework for Web Artisans', ['Laravel', 'Framework', 'Artisans'], 'b');
// <b>Laravel</b>, The PHP <b>Framework</b> for Web <b>Artisans</b>
```
If no `$highlight` or an empty `$tag` is given, the string is returned as-is, because there is nothing to highlight
```php
Str::highlight('Laravel, The PHP Framework for Web Artisans', '');
Str::highlight('Laravel, The PHP Framework for Web Artisans', []);
Str::highlight('Laravel, The PHP Framework for Web Artisans', 'Laravel', '');
// Laravel, The PHP Framework for Web Artisans
```

BramVanBerkel committed:
Add Str::highlight()

BramVanBerkel committed:
formatting

taylorotwell commented:
Hey there - is this the exact same implementation as Rails? What if the string to highlight contains some characters that have meaning in regex?

JayBizzle commented:
Yeah, as Taylor says, this wouldn't work...
```php
Str::highlight('Laravel PHP framework?', '?');
// returns null
```

You probably need to use `preg_quote()`

```php
return preg_replace('/(?:'.preg_quote(implode('|', (array) $highlight)).')/', "<$tag>$0</$tag>", $text);
```

BramVanBerkel committed:
preg_quote highlight variables & take special characters into account

BramVanBerkel committed:
formatting

BramVanBerkel committed:
apply same fix for prepending characters

BramVanBerkel committed:
revert previous fix & add more tests

BramVanBerkel committed:
add forward slash character & add tests

BramVanBerkel commented:
Good catch, thanks!

I've fixed it by `preg_quote()`'ing every highlight variable, to prevent it from escaping the `|` when `implode`'ing.

I've also changed the regex to include characters that have meaning in regex, the default word boundary in regex doesn't take them into account. This way it'll be more intuitive to use.
Old:
```php
//old:
Str::highlight('Laravel?', 'Laravel); // '<mark>Laravel</mark>?'
//new:
Str::highlight('Laravel?', 'Laravel); // '<mark>Laravel?</mark>'
```

JayBizzle commented:
> Old:
> 
> ```php
> 
> //old:
> 
> Str::highlight('Laravel?', 'Laravel); // '<mark>Laravel</mark>?'
> 
> //new:
> 
> Str::highlight('Laravel?', 'Laravel); // '<mark>Laravel?</mark>'
> 
> ```

Not sure I would expect that new output 🤔

JayBizzle commented:
Ah, that's better 👍🏻

BramVanBerkel commented:
I have also opened a PR for the docs, for if we want to merge this: https://github.com/laravel/docs/pull/7733

bert-w commented on tests/Support/SupportStrTest.php:
```diff
@@ -202,6 +202,38 @@ public function testStrExcerpt()
+        $this->assertSame('Laravel PHP framework', Str::highlight('Laravel PHP framework', 'Laravél'));
+        $this->assertSame('<mark>Laravél</mark> PHP framework', Str::highlight('Laravél PHP framework', 'Laravél'));
+
+        $this->assertSame('Laravel PHP <mark>framework?</mark>', Str::highlight('Laravel PHP framework?', 'framework?'));
+        $this->assertSame('Laravel PHP framework?', Str::highlight('Laravel PHP framework?', 'framework'));
```
This second line should resolve to `'Laravel PHP <mark>framework</mark>?'` right?

bert-w commented on tests/Support/SupportStrTest.php:
```diff
@@ -202,6 +202,38 @@ public function testStrExcerpt()
+        $this->assertSame('Laravel PHP <mark>framework?</mark>', Str::highlight('Laravel PHP framework?', 'framework?'));
+        $this->assertSame('Laravel PHP framework?', Str::highlight('Laravel PHP framework?', 'framework'));
+        $this->assertSame('<mark>Laravel:</mark> <mark>PHP</mark> <mark>framework!</mark>', Str::highlight('Laravel: PHP framework!', ['Laravel:', 'PHP', 'framework!']));
+        $this->assertSame('Laravel: a PHP framework', Str::highlight('Laravel: a PHP framework', 'Laravel'));
```
`'<mark>Laravel</mark>: a PHP framework'`

bert-w commented on tests/Support/SupportStrTest.php:
```diff
@@ -202,6 +202,38 @@ public function testStrExcerpt()
+        $this->assertSame('<mark>Laravel:</mark> <mark>PHP</mark> <mark>framework!</mark>', Str::highlight('Laravel: PHP framework!', ['Laravel:', 'PHP', 'framework!']));
+        $this->assertSame('Laravel: a PHP framework', Str::highlight('Laravel: a PHP framework', 'Laravel'));
+        $this->assertSame('<mark>Laravel:</mark> a PHP framework', Str::highlight('Laravel: a PHP framework', 'Laravel:'));
+        $this->assertSame('?Laravel: a PHP framework', Str::highlight('?Laravel: a PHP framework', 'Laravel'));
```
`'?<mark>Laravel</mark>: a PHP framework'`

bert-w commented on tests/Support/SupportStrTest.php:
```diff
@@ -202,6 +202,38 @@ public function testStrExcerpt()
+        $this->assertSame('?Laravel: a PHP framework', Str::highlight('?Laravel: a PHP framework', 'Laravel'));
+        $this->assertSame('<mark>?Laravel:</mark> a PHP framework', Str::highlight('?Laravel: a PHP framework', '?Laravel:'));
+
+        $this->assertSame('Laravel PHP/framework', Str::highlight('Laravel PHP/framework', 'PHP'));
```
`'Laravel <mark>PHP</mark>/framework'`

bert-w reviewed (commented):
These are some unexpected outcomes imho

(P.S. the review refers to the last line in each marked code segment)

JayBizzle commented on src/Illuminate/Support/Str.php:
```diff
@@ -289,6 +289,23 @@ public static function excerpt($text, $phrase = '', $options = [])
+
+        $highlights = array_map(fn ($value) => preg_quote($value, '/'), (array) $highlights);
+
+        return preg_replace('/(?<![\w?])('.implode('|', $highlights).')(?![\w?+:.\/*^$()\[{\\}\]|])/', "<$tag>$0</$tag>", $text);
```
Why are all these characters in the negative look-ahead?

```\w?+:.\/*^$()\[{\\}\]|```

BramVanBerkel replied:
They ensure that for example `Laravel?` doesn't match with `Laravel`, but only with `Laravel?`. Imho this way it's more intuitive use.

JayBizzle replied:
I agree with your example, but it has the undesired effect (in my opinion) of the following output

```php
Str::highlight('Laravel?', 'Laravel');
// Actual output: Laravel?
// Expected output: <mark>Laravel</mark>?
```

JayBizzle replied:
> They ensure that for example `Laravel?` doesn't match with `Laravel`, but only with `Laravel?`. Imho this way it's more intuitive use.

Also, i'm pretty sure it wouldn't actually do this, as the `?` is passed through `preg_quote`

JayBizzle replied:
I think your original implementation was better (and simpler), but just needed to add the `preg_quote`

```php
$highlights = array_map(fn ($value) => preg_quote($value, '/'), (array) $highlights);

return preg_replace('/(?:'.implode('|', $highlights).')/', "<$tag>$0</$tag>", $text);
```

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

BrandonSurowiec commented:
I made this as a macro. It's case insensitive `/i` to match the browser's native "Find" functionality for my use case.

```php
        /**
         * Highlight the given words in a string.
         * 
         * @param  string  $text
         * @param  string|array  $highlights
         * @param  string  $tag
         * @return string
         */
        Str::macro('highlight', function($text, $highlights, $tag = 'mark') {
            if (empty($highlights) || $tag === '') {
                return $text;
            }

            $highlights = array_map(fn ($value) => preg_quote($value, '/'), (array) $highlights);

            return preg_replace('/('.implode('|', $highlights).')/i', "<$tag>$0</$tag>", $text);
        });
```
